### PR TITLE
Exclude its subtree

### PR DESCRIPTION
**Pull request by mtsmfm**

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -12,7 +12,12 @@ class Category < ActiveRecord::Base
   # For PostgreSQL, doesn't work on MySQL
   # See https://github.com/stefankroes/ancestry/pull/238
   scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
-  scope :parental, ->(category = nil) { includes(:posts).where("posts.id" => nil).where.not(id: category&.id).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
+  scope :parental, ->(category) {
+    categories = includes(:posts).where("posts.id" => nil).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order)
+    categories = categories.where.not(id: category.subtree.ids) if category.persisted?
```
Comment (okkez) [inline app/models/category.rb]: How about `subtree_ids`?

File: test/integration/editor_test.rb [hunk 2]
```diff
@@ -423,6 +423,38 @@ class EditorTest < ActionDispatch::IntegrationTest
...
+        within(:row, @categories[3].name) do
+          click_on("Edit")
+        end
+
+        assert_equal(
+          [
+            "",
+            @categories[0].name,
+            [@categories[0], @categories[1]].map(&:name).join("/"),
+            [@categories[0], @categories[1], @categories[2]].map(&:name).join("/")
```
Comment (okkez) [inline test/integration/editor_test.rb]: How about `@categories[2].full_name` ?

File: test/integration/editor_test.rb [hunk 3]
```diff
@@ -423,6 +423,38 @@ class EditorTest < ActionDispatch::IntegrationTest
...
       end
     end
+
+    sub_test_case "parent category" do
+      setup do
+        # 0
+        # └── 1
+        #     ├── 2
+        #     └── 3
+        #         └── 4
```
Comment (okkez) [inline test/integration/editor_test.rb:435]: :+1:

File: test/integration/editor_test.rb (hunk 2 above)
Comment (mtsmfm) [inline test/integration/editor_test.rb]: THANK YOU!

File: app/models/category.rb [hunk 4]
```diff
@@ -12,7 +12,12 @@ class Category < ActiveRecord::Base
   # For PostgreSQL, doesn't work on MySQL
   # See https://github.com/stefankroes/ancestry/pull/238
   scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
-  scope :parental, ->(category = nil) { includes(:posts).where("posts.id" => nil).where.not(id: category&.id).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
+  scope :parental, ->(category) {
```
Comment (mtsmfm) [inline app/models/category.rb]: hi

File: app/models/category.rb [hunk 5]
```diff
@@ -12,7 +12,12 @@ class Category < ActiveRecord::Base
...
+    categories = includes(:posts).where("posts.id" => nil).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order)
```
Comment (mtsmfm) [inline app/models/category.rb:16]: It seems that there is no way to inject

**Comment by okkez** [comment]
LGTM :+1:

File: app/models/category.rb [hunk 6]
```diff
@@ -12,7 +12,12 @@ class Category < ActiveRecord::Base
   # For PostgreSQL, doesn't work on MySQL
   # See https://github.com/stefankroes/ancestry/pull/238
   scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
-  scope :parental, ->(category = nil) { includes(:posts).where("posts.id" => nil).where.not(id: category&.id).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order) }
+  scope :parental, lambda {|category|
+    categories = includes(:posts).where("posts.id" => nil).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order)
+    categories = categories.where.not(id: category.subtree.ids) if category.persisted?
```
Comment (tricknotes) [inline app/models/category.rb]: Using `category.subtree.select(:id).to_sql` reduces count of SQL query.

**Comment by tricknotes** [comment]
I left a comment however totally looks awesome! 🌟

File: app/models/category.rb (hunk 6 above)
Comment (mtsmfm) [inline app/models/category.rb]: thank you!

**Comment by tricknotes** [comment]
🐳

**Comment by tricknotes** [comment]
It seems the test on Circle CI is all green 🍏 
https://circleci.com/gh/bm-sms/daimon-news/2082

**Pull request merged**